pkg/oc/cli/admin/mustgather/mustgather.go

```diff
@@ -0,0 +1,220 @@
+package mustgather
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/spf13/cobra"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/tools/clientcmd/api/latest"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/templates"
+	kcmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+)
+
+var (
+	mustGatherLong = templates.LongDesc(`
+		Launch a pod to gather debugging information
+
+		This command will launch a pod in a temporary namespace on your
+		cluster that gathers debugging information, using a copy of the active 
+		client config context, and then downloads the gathered information. 
+	`)
+
+	mustGatherExample = templates.Examples(`
+		# gather default information using the default image and command, writing into ./must-gather.local.<rand>
+		  oc adm must-gather
+
+		# gather default information with a specific local folder to copy to
+		  oc adm must-gather --dest-dir=/local/directory
+
+		# gather default information using a specific image, command, and pod-dir
+		  oc adm must-gather --image=my/image:tag --source-dir=/pod/directory -- myspecial-command.sh
+	`)
+)
+
+type MustGatherFlags struct {
+	ConfigFlags *genericclioptions.ConfigFlags
+
+	NodeName string
+	DestDir  string
+	Image    string
+
+	genericclioptions.IOStreams
+}
+
+func NewMustGatherFlags(streams genericclioptions.IOStreams) *MustGatherFlags {
+	return &MustGatherFlags{
+		ConfigFlags: genericclioptions.NewConfigFlags(),
+		IOStreams:   streams,
+	}
+}
+
+func NewMustGatherCommand(restClientGetter genericclioptions.RESTClientGetter, streams genericclioptions.IOStreams) *cobra.Command {
+	f := NewMustGatherFlags(streams)
+	cmd := &cobra.Command{
+		Use:     "must-gather",
+		Short:   "Launch a new instance of a pod for gathering debug information",
+		Long:    mustGatherLong,
+		Example: mustGatherExample,
+		Run: func(cmd *cobra.Command, args []string) {
+			kcmdutil.CheckErr(f.Complete(cmd, restClientGetter, args).RunMustGather())
+		},
+	}
+
+	cmd.Flags().StringVar(&f.NodeName, "node-name", f.NodeName, "Set a specific node to use - by default a random master will be used")
+	cmd.Flags().StringVar(&f.Image, "image", f.Image, "Set a specific to use - by default the image will be looked up for OpenShift's must-gather")
+	cmd.Flags().StringVar(&f.DestDir, "dest-dir", f.DestDir, "Set a specific directory on the local machine to write gathered data to.")
+
+	return cmd
+}
+
+func (f *MustGatherFlags) Complete(cmd *cobra.Command, restClientGetter genericclioptions.RESTClientGetter, args []string) *MustGatherOptions {
+	o := &MustGatherOptions{
```

Comment:
Something smells here. You should have only one `MustGatherOptions` struct and work from there through `Complete` -> `Validate` (if needed) -> `Run`. That is the required flow for all commands.

Reply:
Done